modules/phase_field/src/userobjects/EBSDReader.C

```diff
@@ -186,19 +192,126 @@ EBSDReader::readFile()
     a._p += j._p;
     a._n++;
   }
-
-  for (unsigned int i = 0; i < _grain_num; ++i)
+  // creating a vector of maps to store the bin name and its size
+  std::vector<
+      std::map<std::tuple<unsigned int, unsigned int, unsigned int, unsigned int>, unsigned int>>
+      BinToSize;
+  // creating a vector of vectors to store the bin name corresponding to a quaternion
+  std::vector<std::vector<std::tuple<unsigned int, unsigned int, unsigned int, unsigned int>>>
+      QuatToBin(_grain_num);
+  // throws an error for zero bins
+  if (_bins == 0)
+    mooseError("One cannot have zero bins");
+  for (unsigned int i = 0; i < _grain_num; i++)
   {
     EBSDAvgData & a = _avg_data[i];
     EulerAngles & b = _avg_angles[i];
 
     if (a._n == 0)
       continue;
 
-    // TODO: need better way to average angles
-    b.phi1 /= Real(a._n);
-    b.Phi /= Real(a._n);
-    b.phi2 /= Real(a._n);
+    Real w_index, x_index, y_index, z_index;
+    Real w, x, y, z;
+    std::vector<std::tuple<double, Eigen::VectorXd>> eigen_vectors_and_values;
+
+    // creating an empty map
+    BinToSize.push_back(std::map<std::tuple<unsigned int, unsigned int, unsigned int, unsigned int>,
+                                 unsigned int>());
+
+    // looping through quaternions of each grain
+    for (unsigned int k = 0; k < Quat[i].size(); k++)
+    {
+      std::tuple<unsigned int, unsigned int, unsigned int, unsigned int> Bin;
+      w_index = int((Quat[i][k].w() + 1) * 0.5 * _bins);
+      x_index = int((Quat[i][k].x() + 1) * 0.5 * _bins);
+      y_index = int((Quat[i][k].y() + 1) * 0.5 * _bins);
+      z_index = int((Quat[i][k].z() + 1) * 0.5 * _bins);
+
+      Bin = std::make_tuple(w_index, x_index, y_index, z_index);
+
+      // storing the bin corresponding to the quaternion Quat[i][k]
+      QuatToBin[i].push_back(Bin);
+
+      // storing the bin name and its corresponding size value
+      if (BinToSize[i].find(Bin) == BinToSize[i].end())
+        BinToSize[i].insert(std::make_pair(Bin, 1));
+      else
+        BinToSize[i][Bin] += 1;
```

Comment:
There is a recommended way of doing this. It would go:

```
  auto i = BinToSize[i].find(Bin);
  if (i == BinToSize[i].end())
    BinToSize[i].emplace_hint(i, Bin, 1);
  else
    i->second++;
```

This code will only ever perform one map lookup as opposed to your version which performs 2! (please replace the variable names with all lower case and underscore separated names as suggested above)